### ISSUE TITLE
TS allows adding Uint8Array item to Array<ArrayBuffer>

### ISSUE THREAD

**Issue by lll000111** [microsoft/TypeScript#31311]
**Search Terms:** ArrayBuffer, Uint8Array

## Code

[Playground link](http://www.typescriptlang.org/play/index.html#src=interface%20SimpleReadStream%20%7BonData%3A%20(data%3A%20ArrayBuffer)%20%3D%3E%20any%3B%7D%0D%0Adeclare%20function%20createFileReadStream(name%3A%20string)%3A%20SimpleReadStream%3B%0D%0A%0D%0A%2F%2F%20CASE%201%0D%0Aconst%20stream1%3A%20SimpleReadStream%20%3D%20createFileReadStream(name)%3B%0D%0Aconst%20blobData1%3A%20ArrayBuffer%5B%5D%20%3D%20%5B%5D%3B%0D%0Astream1.onData%20%3D%20data%20%3D%3E%20%7B%0D%0A%20%20%20%20%2F%2F%20TYPE%20MISMATCH%0D%0A%20%20%20%20blobData1.push(new%20Uint8Array(data))%3B%0D%0A%7D%3B%0D%0A%0D%0A%2F%2F%20CASE%202%0D%0Aconst%20stream2%3A%20SimpleReadStream%20%3D%20createFileReadStream(name)%3B%0D%0Aconst%20blobData2%3A%20Uint8Array%5B%5D%20%3D%20%5B%5D%3B%0D%0Astream2.onData%20%3D%20data%20%3D%3E%20%7B%0D%0A%20%20%20%20%2F%2F%20TYPE%20MISMATCH%0D%0A%20%20%20%20blobData2.push(data)%3B%0D%0A%7D%3B%0D%0A)

```ts
interface SimpleReadStream {onData: (data: ArrayBuffer) => any;}
declare function createFileReadStream(name: string): SimpleReadStream;

// CASE 1
const stream1: SimpleReadStream = createFileReadStream(name);
const blobData1: ArrayBuffer[] = [];
stream1.onData = data => {
    // TYPE MISMATCH (but no error)
    blobData1.push(new Uint8Array(data));
};

// CASE 2
const stream2: SimpleReadStream = createFileReadStream(name);
const blobData2: Uint8Array[] = [];
stream2.onData = data => {
    // TYPE MISMATCH (error as expected)
    blobData2.push(data);
};
```

**Expected behavior:**

Both test cases should result in type errors.

**Actual behavior:**

Only case 2 results in an error, case 1 does not.

**Comment by nmain** [comment]
Sure, `Array.prototype.push` has covariant inputs, and an instance of `Uint8Array` has all of the properties that an instance of `ArrayBuffer` has, so typescript's structural typing considers `Uint8Array extends ArrayBuffer` to be true.

**Comment by lll000111** [comment]
@nmain But a Uint8Array is not an ArrayBuffer, two very different objects. You can't mix them.

**Comment by nmain** [comment]
Read up on structural typing https://www.typescriptlang.org/docs/handbook/type-compatibility.html, and then wrap your head around this example:

http://www.typescriptlang.org/play/#src=const%20Nmain_SuperArray%20%3D%20%7B%0D%0A%20%20%20%20byteLength%3A%2042%2C%0D%0A%20%20%20%20slice()%20%7B%0D%0A%20%20%20%20%20%20%20%20return%20Nmain_SuperArray%0D%0A%20%20%20%20%7D%2C%0D%0A%20%20%20%20%5BSymbol.toStringTag%5D%3A%20%22Awesome!%22%0D%0A%7D%0D%0A%0D%0A%2F%2F%20passes%20typecheck%0D%0Aconst%20s%3A%20ArrayBuffer%20%3D%20Nmain_SuperArray%0D%0A

In short, all it matters is that `Uint8Array` has `byteLength` and `slice()` that work similarly to `ArrayBuffer`, and it does.

**Comment by lll000111** [comment]
> Read up on structural typing 

I don't need to, I know what it is.

**Comment by nmain** [comment]
I'm sorry my example wasn't clear enough for you.  Let me show you the exact definition for `ArrayBuffer` inside Typescript:

```ts
interface ArrayBuffer {
    readonly byteLength: number;
    slice(begin: number, end?: number): ArrayBuffer;
}
```
(Leaving off `toStringTag` here as not all editions have it, but it doesn't fundamentally change the example.)

Any object that has these methods and properties is compatible with an ArrayBuffer.  It doesn't matter if there's some behavioral difference between them, because TypeScript does not measure those things and TypeScript does not use nominal typing.

In the example I posted, I give an object which is obviously not an ArrayBuffer - `byteLength` is always 42, `slice` returns the same object again regardless of parameters.  But it passes because all Typescript cares is:

1. Does this object have a `byteLength` property of type `number`?
2. Does this object have a `slice()` method, optionally taking 0-2 `number` parameters, that returns an object that also satisfies 1 and 2?

**Comment by lll000111** [comment]
@nmain 

???

https://github.com/microsoft/TypeScript/issues/31311#issuecomment-490581595

**Comment by nmain** [comment]
I am sorry that I misread you.  Your terse comments did not give me much to work with to understand your intent. 

This is a duplicate of https://github.com/Microsoft/TypeScript/issues/202 and there have been other discussions of it as well.

https://github.com/microsoft/TypeScript/issues/10882
https://github.com/microsoft/TypeScript/issues/5855
etc

**Comment by weswigham** [comment]
FYI, in your own project, you can make the various `Buffer` types distinguishable by augmenting them with type-specific tags, eg:
```ts
interface ArrayBuffer {
  " buffer_kind"?: "array"; 
}
interface Uint8Array {
  " buffer_kind"?: "uint8"; 
}
```
we briefly trialed using `Symbol.toStringTag` in this way, but making the buffers distinguishable actually breaks a number of people, apparently.
